jupyterlab/shim-maker.js

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+var path = require('path');
+var fs = require('fs');
+
+
+/**
+ * Generate a shim for a library that does not have top level indexes.
+ *
+ * @param modName (string) - The name of the module to shim.
+ *
+ * @param sourceFolder (string) - The source folder.
+ *
+ * @param regex - A regular expresion used to match files - the default
+ *  is all `.js` files.
+ *
+ * @returns A promise that resolves when the file is created.
+ */
+function shimmer(modName, sourceFolder, regex) {
+  var modPath = require.resolve(modName + '/package.json');
+  sourceFolder = sourceFolder || 'lib';
+  regex = regex || /.*\.js$/;
+  modPath = path.join(path.dirname(modPath), sourceFolder);
+  var lines = ['var ' + modName + ' = {}'];
+
+  var entries = fs.readdirSync(modPath);
+  // Algorithm:
+  // search for index.js file
+  // if one exists, use it, otherwise use an empty object to initialize
+  // then, for each, if it is a directory, recurse-and-add
+  // if it is a file that matches our regex, add it
+  var lines = getLines(modName, sourceFolder, modPath, modPath, regex);
+  lines.push('module.exports = ' + modName);
+  return lines.join('\n');
+}
+
+
+/**
+ * Get the appropriate shim lines for the items in a folder.
+ */
+function getLines(modName, sourceFolder, basePath, currentPath, regex) {
+  var lines = [];
+  var entries = fs.readdirSync(currentPath);
+  var relPath = path.relative(basePath, currentPath);
+  var parts = relPath.split('/');
+  var modPath = path.join(modName, sourceFolder, relPath);
+
+  // Initialize with the index file if available, else an empty object.
+  if (entries.indexOf('index.js') !== -1) {
+    if (!parts[0]) {
+      lines.push('var ' + modName + ' = require("' + modPath + '");');
+    } else {
+      lines.push(modName + '["' + parts.join('"]["') + '"] = require("' + modPath + '");');
+    }
+  } else {
+    if (!parts[0]) {
+      lines.push('var ' + modName + ' = {};');
+    } else {
+      lines.push(modName + '["' + parts.join('"]["') + '"] = {};');
+    }
+  }
+  // Otherwise, handle each file.
+  for (var i = 0; i < entries.length; i++) {
+    var entry = entries[i];
+    // Handle directories recursively.
```

Comment:
For the next iteration - the fs-extra module, which we already require, has a `walk` function.